[FLINK-8756][Client] Support ClusterClient.getAccumulators() in RestClusterClient

yanghua opened the pull request:
## What is the purpose of the change

This PR Support ClusterClient.getAccumulators() in RestClusterClient.

## Brief change log

  - *Send REST request to get `JobAccumulatorsInfo` object*
  - *Use jackson's ObjectMapper convert `JobAccumulatorsInfo` object to Map*
  - *Add a test method into `RestClusterClientTest` class to test the `getAccumulators` function*
  - *Add a test handler to mock `JobAccumulatorsInfo` object*


## Verifying this change

This change added tests and can be verified as follows:

  - *Added test that validates that actual accumulators size equals we mocked in the test handler*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

yanghua commented:
HI @tillrohrmann and @aljoscha , who would review this PR? Thanks!

aljoscha commented:
@yanghua I'll have a look next week. From a first look it seems like we might have to extend the rest handler and the client to also return the accumulators as a `SerializedValue<>` as the old code for accumulators does, to allow user defined accumulator types.

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -363,6 +367,27 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+			params
+		);
+
+		return responseFuture.thenApply((accumulatorsInfo) -> {
```
Adding types to lambdas is always a good idea.

yanghua commented:
@aljoscha and @tillrohrmann , Thanks for your suggestion. I will rework for this. 
Moreover, @tillrohrmann would please review my another PR [FLINK-8459](https://github.com/apache/flink/pull/5565)? Thanks!

yanghua commented:
Hi @aljoscha and @tillrohrmann  I have tried to put `Map<String, SerializedValue<Object>>` into `JobAccumulatorsInfo` with two ways : 
*  a `@JsonProperty` 
* marked with `@JsonIgnore` and provide getter and setter

and the two ways both failed with different reason: 

* first way : throw a exception : `SerializedValue` no suitable constructor;
* second way : the setted object will be ignore by `HandlerUtils#sendResponse`

In `JobAccumulatorsHandler#handleRequest` I can not return `Map<String, SerializedValue<Object>>` directly, because this method's result type must extends `ResponseBody`.

GJL commented:
@yanghua Take a look at `SerializedValueDeserializer` and `SerializedValueSerializer`. You can use `@JsonSerialize(using= ... )`, `@JsonSerialize(keyUsing= ... )`, and `@JsonDeserialize` to specify the class for ser/des.

yanghua commented:
@GJL thanks for your guidance! 
 @aljoscha  and @tillrohrmann I have refactored this issue, remained the `getAccumulators` method created a `getSerializedAccumulators` method to return the serialized accumulators in `ClusterClient`. And override `getSerializedAccumulators` in `RestClusterClient`. For `JobAccumulatorsInfo`, I also remained `UserTaskAccumulator` and just added a new property for serialized accumulators. Please review, thanks.

aljoscha commented:
I think it would be good to be able to query both representations. The stringified version can be good if you want to do a quick query using a rest client. And we don't have to change the Web UI if we keep both versions.

yanghua commented:
@aljoscha yes, I agree. 

> I think it would be good to be able to **query** both representations

The "query" you mean `JobAccumulatorsHandler` or `RestClusterClient` behavior?

In `JobAccumulatorsHandler#handleRequest` method, we could query the accumulator's string and serialized representations, and boxed in `JobAccumulatorsInfo` object. 

1. let `getAccumulators` return accumulator's string representations and `getSerializedAccumulators` method return accumulator's serialized representations.

2.  let `getAccumulators` return both representations?

Which one is your idea?

aljoscha commented:
@yanghua I mean having two separate ways of querying: the server should have two endpoints (or differentiate based on a query parameter) and return either the stringified accumulators or the serialized accumulators. For now, the former would be used for the web frontend while the latter would be used by the `RestClusterClient`. @GJL maybe has a better opinion on how this should be done on the server side, though.

yanghua commented:
OK, base of @aljoscha 's opinion, I describe the implementation detail before coding : 

* define a `AccumulatorRepresentationsQueryParameter` class which extends `MessageQueryParameter` and contains to two representation mode (***stringified***  and ***serialized***)

* define a `JobAccumulatorsMessageParameters` class which extends `MessageParameters`, and the `getQueryParameters` method will return the `AccumulatorRepresentationsQueryParameter`

* refactor `JobAccumulatorsHeaders#getUnresolvedMessageParameters` return `JobAccumulatorsMessageParameters`'s instance

* refactor `JobAccumulatorsHandler#handleRequest` it will query specific accumulator's representation base of `AccumulatorRepresentationsQueryParameter`, and the `JobAccumulatorsInfo` will be reused for both representations

* in `RestClusterClient` class , let `getAccumulators` return stringified accumulators(`Map<String, Object>`) and `getSerializedAccumulators` return serialized accumulators(`Map<String, SerializedValue<Object>>`)

@tillrohrmann and @GJL  hope for your opinions, Thanks.

aljoscha commented:
I can't comment on the first points but on the last point. I would suggest to only add `getAccumulators()` to `RestClusterClient` that behaves the same way as the existing `getAccumulators()` on `ClusterClient`: it retrieves the serialised values and returns a map of the deserialised values.

GJL commented:
I agree with @aljoscha. We should not add more `public` methods to `ClusterClient`. Implementing `getAccumulators()` in `RestClusterClient` is enough. 

Regarding the `JobAccumulatorsHandler`, what would speak against adding a new field to `UserTaskAccumulator` which stores the `SerializedValue`? The JSON representation would always carry an additional base64 encoded `byte` array but I think performance isn't important at this point.

cc: @tillrohrmann @aljoscha

yanghua commented:
@GJL , the before discussion between I and @aljoscha treat the accumulators as a whole (maybe only me?). Because the pre-existing implementation `getAccumulatorResultsStringified ` and `getAccumulatorsSerialized ` , so I took steps along this path and split the stringified and serialized accumulators.

So the conclusion is : accepting your opinion and taking the each key's SerializedValue into `UserTaskAccumulator` ? 

If yes, I will try to change the code.

GJL commented:
@yanghua 
I talked to @tillrohrmann offline, and we decided it is enough to add a query parameter such as `includeSerializedValue` to the `JobAccumulatorsHandler`. If `includeSerializedValue` is `true`, then `SerializedValue` should be part of the JSON response (in addition to the stringified value), otherwise only the stringified value. By default `includeSerializedValue` should be `false` because the Web UI cannot handle binary data.  For the request in `RestClusterClient` you would always set the flag to `true`.

Let me know if you have any questions.

cc: @tillrohrmann

yanghua commented:
@GJL I have refactored the code and test case, could you please review the new commit? Thanks.

GJL commented:
@yanghua I will take a look this week.

yanghua commented:
@tillrohrmann , if you have time could you please review this PR? Thanks.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobAccumulatorsQueryParameter.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * query parameter for job's accumulator handler {@link org.apache.flink.runtime.rest.handler.job.JobAccumulatorsHandler}.
+ */
+public class JobAccumulatorsQueryParameter extends MessageQueryParameter<String> {
```
Why not directly make this a boolean option?

zentol replied:
The name is also a bit generic.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobAccumulatorsInfo.java:
```diff
@@ -104,6 +129,18 @@ public UserTaskAccumulator(
 			this.value = Preconditions.checkNotNull(value);
 		}
 
+		public String getName() {
```
missing `@JsonIgnore` annotation (also applies to other getters

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobAccumulatorsInfo.java:
```diff
@@ -33,19 +39,38 @@
+		this.serializedUserAccumulators = Preconditions.checkNotNull(serializedUserAccumulators);
+	}
+
+	public List<JobAccumulator> getJobAccumulators() {
```
missing `@JsonIgnore` annotations

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +394,33 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+			if (accumulatorsInfo != null) {
+				Map<String, Object> result = new HashMap<>(3);
+
+				result.put(JobAccumulatorsInfo.FIELD_NAME_JOB_ACCUMULATORS, accumulatorsInfo.getJobAccumulators());
```
The resulting API (as shown in `RestClusterClientTest`) is effectively not usable for users and inconsistent with existing behavior in `ClusterClient`.

The returned map should only contain the deserialized accumulators with their respective name as the key.

zentol commented:
We may want to delay merging this until [FLINK-8881](https://issues.apache.org/jira/browse/FLINK-8881) has been addressed, as it voids the primary use-case of this handler.

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +393,27 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
 	}
 
+	@Override
+	public Map<String, Object> getAccumulators(final JobID jobID) throws Exception {
```
we should also override `Map<String, Object> getAccumulators(JobID jobID, ClassLoader loader)`

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +393,27 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+
+		return responseFuture.thenApply((JobAccumulatorsInfo accumulatorsInfo) -> {
+			if (accumulatorsInfo != null && accumulatorsInfo.getSerializedUserAccumulators() != null) {
+				return accumulatorsInfo.getSerializedUserAccumulators();
```
the accumulators should be deserialized via `SerializedValue#deserialize(ClassLoader)` .

If `Map<String, Object> getAccumulators(JobID jobID, ClassLoader loader)` (that also should be overridden) was called use the passed in `ClassLoader`, otherwise `ClassLoader.getSystemClassLoader()`.

yanghua commented:
hi @zentol , I refactor the code based on your review suggestion. Would you please review again, thanks!

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobAccumulatorsMessageParameters.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class JobAccumulatorsMessageParameters extends JobMessageParameters {
+
+	public final AccumulatorsIncludeSerializedValueQueryParameter queryParameter = new AccumulatorsIncludeSerializedValueQueryParameter();
```
field name is a bit generic; how about `includeSerializedAccumulators`?

yanghua commented:
@zentol it seems the Travis CI has some problem, always build failed. Please review my latest change.

zentol commented:
@yanghua This test failure is unrelated, you can ignore it.

GJL commented:
You can use `git commit -m "Trigger build" --allow-empty` to trigger a build.

zentol commented:
please do not trigger builds just to get that perfect green build. The test failure here is quite common at the moment.

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +394,36 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+	@Override
+	public Map<String, Object> getAccumulators(JobID jobID) throws Exception {
+		return getAccumulators(jobID, ClassLoader.getSystemClassLoader());
+	}
```
Should not be necessary to override.

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +394,36 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+				try {
+					return AccumulatorHelper.deserializeAccumulators(accumulatorsInfo.getSerializedUserAccumulators(), loader);
+				} catch (Exception e) {
+					log.error("Deserialize accumulators with customized classloader error : {}", e);
```
Let us properly fail with an exception here.

tillrohrmann replied:
Moreover, when logging an exception you don't have to specify a `{}` placeholder.

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +394,36 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+				}
+			}
+
+			return Collections.EMPTY_MAP;
```
please replace with `Collections.<String, Object>emptyMap()`

tillrohrmann commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java:
```diff
@@ -542,6 +549,68 @@ public void testListJobs() throws Exception {
+		}
+	}
+
+	private class TestAccumulatorHandlers  {
```
Why having this extra wrapping class?

yanghua replied:
this follows the `TestSavepointHandlers` in `RestClusterClientTest`, I will remove it.

tillrohrmann commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java:
```diff
@@ -542,6 +549,68 @@ public void testListJobs() throws Exception {
+
+			@Override
+			protected CompletableFuture<JobAccumulatorsInfo> handleRequest(@Nonnull HandlerRequest<EmptyRequestBody,
+				JobAccumulatorsMessageParameters> request, @Nonnull DispatcherGateway gateway) throws RestHandlerException {
```
parameter breaking not consistent

tillrohrmann commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java:
```diff
@@ -542,6 +549,68 @@ public void testListJobs() throws Exception {
+				JobAccumulatorsInfo accumulatorsInfo;
+				List<Boolean> queryParams = request.getQueryParameter(AccumulatorsIncludeSerializedValueQueryParameter.class);
+
+				boolean includeSerializedValue = false;
```
let's make it final and assign false in the else branch

tillrohrmann commented on flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java:
```diff
@@ -542,6 +549,68 @@ public void testListJobs() throws Exception {
+					includeSerializedValue = queryParams.get(0);
+				}
+
+				List<JobAccumulatorsInfo.UserTaskAccumulator> userTaskAccumulators = new ArrayList<JobAccumulatorsInfo.UserTaskAccumulator>() {{
```
Instead of creating a new anonymous class can we simply create an `ArrayList` and then add the element (without the initializer block)?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobAccumulatorsHandler.java:
```diff
@@ -62,18 +64,33 @@ public JobAccumulatorsHandler(
+		List<Boolean> queryParams = request.getQueryParameter(AccumulatorsIncludeSerializedValueQueryParameter.class);
 
-		for (StringifiedAccumulatorResult acc : accs) {
+		boolean includeSerializedValue = false;
```
let's make it `final` and assign `false` in the else branch.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/AccumulatorsIncludeSerializedValueQueryParameter.java:
```diff
@@ -0,0 +1,41 @@
+package org.apache.flink.runtime.rest.messages;
+
+/**
+ * query parameter for job's accumulator handler {@link org.apache.flink.runtime.rest.handler.job.JobAccumulatorsHandler}.
```
Please start with a capital letter.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobAccumulatorsMessageParameters.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.Collections;
+
+/**
+ * request parameter for job accumulator's handler {@link org.apache.flink.runtime.rest.handler.job.JobAccumulatorsHandler}.
```
Capital letter

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/json/SerializedValueDeserializer.java:
```diff
@@ -34,6 +36,10 @@
 
 	private static final long serialVersionUID = 1L;
 
+	public SerializedValueDeserializer() {
```
Why do we add this default constructor? I prefer to specify the type explicitly. Also if it is a simple `TypeReference<SerializedValue<Object>>`.

yanghua replied:
this is because the `SerializedValueDeserializer ` would be used in a jackson annotation `@JsonDeserialize`

tillrohrmann replied:
Alright, this makes sense.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/json/SerializedValueSerializer.java:
```diff
@@ -36,6 +38,10 @@
 	private static final long serialVersionUID = 1L;
 
+	public SerializedValueSerializer() {
+		super(TypeFactory.defaultInstance().constructType(new TypeReference<SerializedValue<Object>>() {}));
```
Same here with the default constructor.

tillrohrmann commented on flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java:
```diff
@@ -389,6 +394,36 @@ public String cancelWithSavepoint(JobID jobId, @Nullable String savepointDirecto
+		);
+
+		return responseFuture.thenApply((JobAccumulatorsInfo accumulatorsInfo) -> {
+			if (accumulatorsInfo != null && accumulatorsInfo.getSerializedUserAccumulators() != null) {
```
`accumulatorsInfo` should always be null. The same applies to `getSerializedUserAccumulators`. Thus there is no need for a null check.

yanghua commented:
hi @tillrohrmann it seems I make a wrong git operation, I merged some new commits and squashed into one, then pushed into the PR(branch). What should I do to fix this problem?

tillrohrmann commented:
Not sure what exactly happened but you can always go back to an earlier
point via the reflog and then a git hard reset.

On Tue, Mar 20, 2018 at 4:11 PM, vinoyang <notifications@github.com> wrote:

> hi @tillrohrmann <https://github.com/tillrohrmann> it seems I make a
> wrong git operation, I merged some new commits and squashed into one, then
> pushed into the PR(branch). What should I do to fix this problem?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink/pull/5573#issuecomment-374634057>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFfXumKiEJDwdbPtTNYflKOUF_6MbW3Cks5tgRwOgaJpZM4SRyVO>
> .
>

yanghua commented:
@tillrohrmann thanks for your reply, it's a misoperation, I have done rollback and refactored the code based on your suggestion. Please check the change again.

tillrohrmann commented:
Changes look good @yanghua. Merging this PR.
